Revamp our documentation and refactor a bit

choldgraf opened the pull request:
This updates our documentation to make it a bit easier to navigate and understand. It also makes some content improvements along the way.

[Docs preview of this PR is here](https://repo2docker--1433.org.readthedocs.build/en/1433/)

Here are some major things that this PR tries to improve:

- The `get started` guide is now a tutorial, meant to be followed step by step, in line with how Diataxis recommends tutorials be used. It now has fewer information, and focuses specifically on "getting started"
- The `howto` section has been renamed the `user guide`, and now groups pages based on major topic area (e.g. `Configuration`, `Deployment`, etc).
- The giant list of configuration files is broken into sub-sections based on the kind of workflow they're attached to (e.g. `datascience`) to make it easier to parse.
- A lot of content has been consolidated and-or moved around to make it easier to find the right information in one place.
- Some updates to styling so that we hide sidebars when they aren't needed, improvements to the dark-mode logo, etc.
- Updating a lot of content to markdown so that it's easier to maintain and improve.
- Uses the `dirhtml` builder for "clean" URLs. If we merge this, we should enable `clean URLs` with readthedocs so that any direct links will properly redirect

More updates

- Adds a little guide on when to use repo2docker in different scenarios.
- Creates a short how-to for each scenario with example code and external links.
- Updates the landing page so that it's more visually appealing and quick to parse, and "feels like" a landing page.

It probably could use a pass or two to double check that the structure works and that I haven't missed something. Happy to iterate with feedback if folks would like.

yuvipanda commented:
@chuckwondo you can see it in https://repo2docker--1433.org.readthedocs.build/en/1433/! That link is accessible via the tests - it's the read the docs

choldgraf commented:
The latest PR adds a bit more content using some slides from a presentation that @yuvipanda has given a few times about repo2docker. Here are some highlights (copied to the PR body as well):

- Adds a little guide on when to use repo2docker in different scenarios.
- Creates a short how-to for each scenario with example code and external links.
- Updates the landing page so that it's more visually appealing and quick to parse, and "feels like" a landing page.

I think this is ready for somebody to take a pass at it - many thanks for the review @chuckwondo - I'll get to your questions when I take another pass after you or somebody else gives it a look.

rgaiacs reviewed (approved):
Great revamp of the documentation. I added minor comments. I'm happy for this to be merged and further improvement to be made in new pull requests.

choldgraf commented on docs/source/configuration/datascience.md:
```diff
@@ -0,0 +1,59 @@
+# Configuration for datascience workflows
```
I'll go with "research workflows"

choldgraf commented:
I believe that I've addressed all of the comments except for a couple. I'm not sure whether we recommend "just" `repo2docker` or the full `jupyter-repo2docker`, I'm also not sure what our policy is on calling them `container images` vs. `Docker images`. Could we resolve those in a follow-up PR? This one is already quite beefy!

rgaiacs commented on docs/source/cli.md:
````diff
@@ -0,0 +1,89 @@
+`repo2docker` is called with this command:
+
+```
+jupyter-repo2docker <source-repository>
````
I don't know what is the "preferred" way. I'm OK with the "preferred" way to be `jupyter-repo2docker` given that the package on PyPI is [`juypter-repo2docker](https://pypi.org/project/jupyter-repo2docker/) and there are a couple of packages with `repo2docker` in their name on PyPI.

<img width="1920" height="1080" alt="image" src="https://github.com/user-attachments/assets/6423d5ae-1252-46e3-aedb-5fe40d57cfe7" />

rgaiacs commented:
Yes, I can send a PR covering the use of "container image".

rgaiacs commented on docs/source/configuration/actions.md:
```diff
@@ -0,0 +1,33 @@
+
+(postbuild)=
+
+## `postBuild` - Run code after installing the environment
```
The `jovyan` user is defined in https://github.com/jupyterhub/repo2docker/blob/5d3a946b24608f1856218eb89861d552e9b826af/repo2docker/app.py#L223-L235 `repo2docker` has no further details about the `jovyan` user. [Someone asked about the `jovyan` user on Jupyter Discourse forum](https://discourse.jupyter.org/t/what-is-with-the-weird-jovyan-user/1673) that includes a link to the section "[What is a Jovyan?](https://docs.jupyter.org/en/latest/community/content-community.html#what-is-a-jovyan)" in the community part of Jupyter documentation.

choldgraf commented:
Feel free to ping me and I'm happy to review. Shall we merge this one and then iterate there?

chuckwondo commented on docs/source/use/community-image.md:
```diff
@@ -0,0 +1,53 @@
+# Use a community-maintained image in a JupyterHub
+
+The simplest way to deploy environment images to a JupyterHub or BinderHub is to _find and re-use a pre-existing image maintained by a community_. In this case, you won't have to maintain anything - you'll simply use a community's image and contribute upstream if you wish.
```
```suggestion
The simplest way to deploy environment images to a JupyterHub or BinderHub is to _find and re-use a pre-existing image maintained by a community_. In this case, you won't have to maintain anything -- you'll simply use a community's image and contribute upstream if you wish.
```

chuckwondo commented on docs/source/use/community-image.md:
```diff
@@ -0,0 +1,53 @@
+
+## First, find a community-maintained image for your workflow
+
+Many communities define their own user envionment images for re-use. Here are a few to look into.
```
```suggestion
Many communities define their own user envionment images for re-use. Here are a few to look into:
```

chuckwondo commented on docs/source/use/community-image.md:
```diff
@@ -0,0 +1,53 @@
+
+Once you have an image and a tag, you can configure your JupyterHub to use it.
+
+Here's sample configuration if you're using the [Zero to JupyterHub for Kubernetes distribution](https://z2jh.jupyter.org).
```
```suggestion
Here's a sample configuration if you're using the [Zero to JupyterHub for Kubernetes distribution](https://z2jh.jupyter.org):
```

chuckwondo commented on docs/source/use/community-image.md:
````diff
@@ -0,0 +1,53 @@
+    tag: 2023.04.15
+```
+
+Here's sample configuration if you're using the [Dockerspawner for JupyterHub spawner](https://jupyterhub-dockerspawner.readthedocs.io/en/latest/).
````
```suggestion
Here's a sample configuration if you're using the [Dockerspawner for JupyterHub spawner](https://jupyterhub-dockerspawner.readthedocs.io/en/latest/):
```

chuckwondo commented on docs/source/use/extend-community-image.md:
````diff
@@ -0,0 +1,85 @@
+    - otter-grader
+```
+
+A repository `Dockerfile` inherits from the upstream image. For example, the following inherits the scipy-notebook image with the `2023-05-01` tag, and copies
````
and copies ??? (missing the rest of the sentence?)

chuckwondo commented on docs/source/use/pathways.md:
````diff
@@ -0,0 +1,100 @@
+
+Many users of `repo2docker` primarily wish to define the environment for a Binder or a JupyterHub. Here are four ways to accomplish this, from least-to-most work.
+
+```{figure} ../_static/images/whentouse.svg
````
I find this figure rather confusing, as there's no obvious start or end point, but making a more intuitive diagram is perhaps for a separate issue. Fortunately, the prose that follows is quite clear.

choldgraf replied:
I'm adding a more standard decision tree structure in the next commit that should clarify it a bit

chuckwondo commented on docs/source/use/repository.md:
```diff
@@ -0,0 +1,63 @@
+
+## Supported repository providers
+
+repo2docker can fetch repositories from a number of repositories. Here are the ones we support:
```
```suggestion
`repo2docker` can fetch repositories from a number of repositories. Here are the various types of _repository references_ we support:
```

chuckwondo commented on docs/source/use/repository.md:
````diff
@@ -0,0 +1,63 @@
+In each case you can build from these repository sources like so:
+
+```bash
+jupyter-repo2docker <URL-ID-or-path>
````
```suggestion
jupyter-repo2docker <repository-reference>
```

choldgraf commented on docs/source/faq.md:
```diff
@@ -0,0 +1,111 @@
+## Can I add executable files to the user's PATH?
+
+Yes! With a {ref}`postBuild` file, you can place any files that should be called
+from the command line in the folder `~/.local/`. This folder will be
```
I think it's `/bin`, I'll add that

choldgraf commented on docs/source/index.md:
````diff
@@ -0,0 +1,72 @@
+When you call `repo2docker` like so:
+
+```
+jupyter-repo2docker <URL of repository>
````
yep good call

choldgraf commented:
Thanks @chuckwondo for the extra comments - I've gone through and made another round of edits to address your suggestions. *Thank you* to both of you for your help.

I'll merge this one in since we have one explicit and one implicit approval, and please ping me if y'all would like to iterate on the content more from here!
